source/slang/slang-options.cpp

```diff
@@ -1635,6 +1644,29 @@ SlangResult OptionsParser::_parseProfile(const CommandLineArg& arg)
     return SLANG_OK;
 }
 
+// Creates a target of the specified type whose output will be embedded as IR metadata
+SlangResult OptionsParser::addEmbeddedLibrary(const CodeGenTarget format, CompilerOptionName option, const char * capability)
+{
+    RawTarget rawTarget;
+    rawTarget.format = format;
+    // Silently allow redundant targets if it is the same as the last specified target.
+    if (m_rawTargets.getCount() == 0 || m_rawTargets.getLast().format != rawTarget.format)
+    {
+        m_rawTargets.add(rawTarget);
+    }
+
+    getCurrentTarget()->optionSet.add(option, true);
+    getCurrentTarget()->optionSet.add(CompilerOptionName::GenerateWholeProgram, true);
+    CapabilityName atom = findCapabilityName(UnownedTerminatedStringSlice(capability));
+    if (atom == CapabilityName::Invalid)
+    {
+        return SLANG_FAIL;
+    }
+    addCapabilityAtom(getCurrentTarget(), atom);
```

Comment:
From what I understand the goal here is to error if capabilities are not at a minimum compatible with `dxil_lib`. 

Assuming this is the goal, what we can do is error when `-embed-dxil` is set and our entry-point is not compatible with the capabilities (we can also check if `-ignoreCapabilities` is turned on, if not on we can just warn and upgrade capabilities instead using `maybeDiagnoseWarningOrError`)

Example below
```
// validateEntryPoint
auto targetCaps = target->getTargetCaps();
auto stageCapabilitySet = CapabilitySet(entryPoint->getProfile().getCapabilityName());
targetCaps.join(stageCapabilitySet);
...
if(target->compilingToDxilLib() && !targetCaps.atLeastOneSetImpliedInOther(CapabilitySet(CapabilityName::dxil_lib)))
{
    sink->diagnose(entryPointFuncDecl, Diagnostics::entryPointIncompatibleWithDxilLib, CapabilitySet(CapabilityName::dxil_lib), targetCaps); // error ...: dxil libraries require $0, entry point compiled with $1.
    /// optionally the code here may be changed such that `maybeDiagnoseWarningOrError` is used. We will upgrade capabilities and warn if  `-ignoreCapabilities` is toggled on.
}
```

Reply:
Thanks, I'll try that tomorrow.  Having slang check that the profile is at least lib_6_3 is what I thought you were suggesting, so it's what I was expecting to happen.

I think a better user experience would be if the minimum profile version was automatically applied for the -embed-dxil target request.  Can the "dxil_lib" capability be applied such that it would actually cause the target request to be 6_3 minimum, or higher if the user explicitly requested a newer version?

Reply:
> Can the "dxil_lib" capability be applied such that it would actually cause the target request to be 6_3 minimum, or higher if the user explicitly requested a newer version?

We can do is something like this:
```
// validateEntryPoint
auto targetCaps = target->getTargetCaps();
auto stageCapabilitySet = CapabilitySet(entryPoint->getProfile().getCapabilityName());
targetCaps.join(stageCapabilitySet);
...
auto dxilLibCapabilitySet = CapabilitySet(CapabilityName::dxil_lib);
if(target->compilingToDxilLib() && !targetCaps.atLeastOneSetImpliedInOther(dxilLibCapabilitySet))
{
     // If `RestrictiveCapabilityCheck` is true we will error, else we will warn.
     // error ...: dxil libraries require $0, entry point compiled with $1.
     // warn ...: dxil libraries require $0, entry point compiled with $1, implicitly upgrading capabilities.
    maybeDiagnoseWarningOrError(
                        sink,
                        target->getOptionSet(),
                        DiagnosticCategory::Capability,
                        entryPointFuncDecl->loc,
                        Diagnostics::entryPointIncompatibleWithDxilLib,
                        Diagnostics::entryPointIncompatibleWithDxilLibRestrictive,
                        dxilLibCapabilitySet,
                        targetCaps);
       
       // add dxil_lib to the cooked targetCaps
      auto targetCaps = target->getTargetCaps();
      targetCaps.join(dxilLibCapabilitySet);
      if(targetCaps.isInvalid())
      {
          sink->diagnose(..., CapabilityName::hlsl, target->getTargetCaps().getCompileTarget()); // error ...: failed to upgrade capabilities, '$0' is incompatible with '$1'
      }
      else
      {
          target->setTargetCaps(targetCaps);
      }
}
```

Note: setting `target->setTargetCaps(targetCaps);` may be the wrong way todo this...

Reply:
Thank you for the code, Ariel! One thing though, validateEntryPoint() would only be applied when targets include an entrypoint. Typically lib targets do not, so I think the validation needs to happen somewhere else.  Perhaps inside precompileForTargets.

Reply:
I see what you are saying, you are correct.
We can do something like this then:
```suggestion
    CapabilityName atom = findCapabilityName(UnownedTerminatedStringSlice(capability));
    if (atom == CapabilityName::Invalid)
    {
        return SLANG_FAIL;
    }
    addCapabilityAtom(getCurrentTarget(), atom);
```
```
// source/slang/slang-compiler-tu.cpp:20
```suggestion
    CapabilityName atom = findCapabilityName(UnownedTerminatedStringSlice(capability));
    if (atom == CapabilityName::Invalid)
    {
        return SLANG_FAIL;
    }
    addCapabilityAtom(getCurrentTarget(), atom);
```
CapabilityName precompileRequirement = CapabilityName::Invalid;
switch (targetReq->getTarget())
{
case CodeGenTarget::DXIL:
    linkage.addTarget(Slang::CodeGenTarget::DXIL);
    precompileRequirement = CapabilityName::dxil_lib;
    break;
default:
    assert(!"Unhandled target");
    break;
}
SLANG_ASSERT(precompileRequirement != CapabilityName::Invalid);
        
auto targetCaps = targetReq->getTargetCaps();
auto dxilLibCapabilitySet = CapabilitySet(CapabilityName::dxil_lib);
if(target->compilingToDxilLib() && !targetCaps.atLeastOneSetImpliedInOther(dxilLibCapabilitySet))
{
     // If `RestrictiveCapabilityCheck` is true we will error, else we will warn.
     // error ...: dxil libraries require $0, entry point compiled with $1.
     // warn ...: dxil libraries require $0, entry point compiled with $1, implicitly upgrading capabilities.
    maybeDiagnoseWarningOrError(
                        sink,
                        target->getOptionSet(),
                        DiagnosticCategory::Capability,
                        entryPointFuncDecl->loc,
                        Diagnostics::entryPointIncompatibleWithDxilLib,
                        Diagnostics::entryPointIncompatibleWithDxilLibRestrictive,
                        dxilLibCapabilitySet,
                        targetCaps);
       
       // add dxil_lib to the cooked targetCaps
      targetCaps.join(dxilLibCapabilitySet);
      if(targetCaps.isInvalid())
      {
          sink->diagnose(..., CapabilityName::hlsl, target->getTargetCaps().getCompileTarget()); // error ...: failed to upgrade capabilities, '$0' is incompatible with '$1'
      }
      else
      {
          targetReq->setTargetCaps(targetCaps);
      }
}
```